add errors into testSteps result

MateuszJeziorski opened the pull request:
Currently if test fails because of javascript error (not assertion), jenkinsCI shows such tests as passed.
Let's pass error message into testSteps result, then we can reuse it to generate junit report that includes (or not) these errors

MateuszJeziorski commented:
@beatfactor any thoughts about this PR? It really blocks us from using nightwatchJS + jenkinsCI

beatfactor commented:
I was thinking actually to make the test fail if there were errors during. But the errors could also be published on the testResults object.

What kind of errors are you running into? If they are selenium errors a better way to handle this would be to fix those errors.

MateuszJeziorski commented:
I have my own JUnitReporter and I've already added
```
var errorsMessages = testcase.errorsMessages;
<% if (errorsMessages && errorsMessages.length) { %>
    <% for (var i = 0; i < errorsMessages.length; i++) { %>
        <error message="<%= errorsMessages[i] %>"></error>
    <% } %>
<% } %>
```
but I didn't want to push it to everyone, since maybe not everyone is interested in such feature.

Errors I'm dealing with are some JS errors. I don't want to add assertion for each single operation, so each JS error I want to treat as test failure. Ex. let's get ```attribute value``` of some element, split it on ```_``` and do something with ```[1]```. If attribute value is incorrect, then ```[1]``` may not exists. It means something went wrong. It is often when I see all tests ```passed``` in jenkins report, but when I check jenkins console log, there are some JS errors so test actually has not passed

beatfactor commented:
But you can also fail the build if the test runner logs an error, can't you? So in this case the test would be marked as passed but the build would still fail. 

I think the reporter change could also be useful if this is to be merged.

beatfactor commented:
Also a test would be nice.

MateuszJeziorski commented:
actually jenkins build fails, because exit code is not 0. But email notification doesn't include tests that failed because of JS errors. Also we have few known bugs, so we expect jenkins job to be failed, but email notification should includes all failing tests
Ok, I can include reported change and add some tests for it today evening

MateuszJeziorski commented:
@beatfactor any update?

beatfactor commented:
Sorry for the lack of progress here, I'll look into it these days.

beatfactor commented:
Again, sorry for the delay. Could you please rebase onto latest master and squash the commits into one?

beatfactor commented on lib/runner/run.js:
```diff
@@ -58,6 +58,7 @@ module.exports = new (function() {
         errors  : results.errors,
         skipped : results.skipped,
         assertions  : [].concat(results.tests),
+        errorsMessages: (Array.isArray(errors) && errors.length) ? errors.slice() : null,
```
`errors` is always an array so this can just be: 'errorsMessages: errors.slice()`.

MateuszJeziorski replied:
few lines below there is a check
```if (Array.isArray(errors) && errors.length) {...}```
I copied&pasted condition just in case.

beatfactor replied:
Oh, I see :-).

gregoryduckworth commented:
This is something I have come across with using Jenkins and would love to see this merged in!
